### PR TITLE
Fix two const_trait_impl issues

### LINKED ISSUE #100222: `#[const_trait]` and super trait error
I tried this code:

```rust
#![feature(const_trait_impl)]

#[const_trait]
pub trait Index {
    type Output;
}

#[const_trait]
pub trait IndexMut: Index {
    fn foo(&mut self) -> <Self as Index>::Output;
}

```

I expected to see this happen: compiles
Instead, this happened: errors

### Meta

latest nightly

### LINKED ISSUE #100543: regression: can't compare `X` with `_` in const contexts
Found in [crater](https://crater-reports.s3.amazonaws.com/beta-1.64-2/index.html):

* https://crater-reports.s3.amazonaws.com/beta-1.64-2/beta-2022-08-09/gh/aharisu.navi/log.txt
   * Looking briefly into the code it's doing something pretty weird (e.g., transmutes [here](https://github.com/aharisu/navi/blob/38558dc0bb326124e273ff73d87f43f970dc4f62/src/value.rs#L11)), but at the very least this diagnostic seems unfortunate.

### PR DESCRIPTION
**Pull request by fee1-dead**
r? @oli-obk 

Fixes #100222.
Fixes #100543.

### CODE REVIEW

File: compiler/rustc_typeck/src/check/wfcheck.rs [hunk 1]
```diff
@@ -1518,12 +1537,19 @@ fn check_fn_or_method<'tcx>(
...
             Some(WellFormedLoc::Param { function: def_id, param_idx: i.try_into().unwrap() }),
             input_ty.into(),
+            None,
         );
     }
 
     implied_bounds.extend(sig.inputs());
 
-    wfcx.register_wf_obligation(hir_decl.output.span(), None, sig.output().into());
+    // override the env when checking the return type. `~const` bounds can be fulfilled with non-const implementations.
```
Comment (compiler-errors) [inline compiler/rustc_typeck/src/check/wfcheck.rs:1546]: Shouldn't you also be applying this to arg types too?

Comment (compiler-errors) [reply compiler/rustc_typeck/src/check/wfcheck.rs:1546]: Also, I guess a deeper question is whether this should be fixed in `rustc_trait_selection::traits::wf` instead, or I guess whether we should be doing wfcheck with a `Constness::Const` param-env at all. (e.g. should we just be doing `param_env.without_const` when constructing the wfcx...?)

Like, is it ever required to register a wf obligation with `Constness::Const` for correctness sake?

Comment (fee1-dead) [reply compiler/rustc_typeck/src/check/wfcheck.rs:1546]: > whether we should be doing wfcheck with a `Constness::Const` param-env at all?

We should. For an `impl` to be well-formed, all super traits must be fulfilled. In the case of `const` super traits, we need to ensure that the parent trait has a `const` `impl`.

Comment (compiler-errors) [reply compiler/rustc_typeck/src/check/wfcheck.rs:1546]: I guess I mean for the nominal obligations of the projection types in the signature, for example. Not necessarily when we verify that a trait's supertraits are satisfied. 

Put in a different way, any time we instantiate a WellFormed predicate, is it ever wrong to pass a param env with no constness there instead?

Comment (fee1-dead) [reply compiler/rustc_typeck/src/check/wfcheck.rs:1546]: That seems desirable. This is related to #100543. I'm going to look into this further.

Comment (estebank) [reply compiler/rustc_typeck/src/check/wfcheck.rs:1546]: I tend to agree with @compiler-errors' reasoning here (it was my first thought when looking at the code).

**Comment by estebank** [comment]
@oli-obk is currently out. CC @rust-lang/types 

r? rust-lang/types

File: compiler/rustc_typeck/src/check/wfcheck.rs (hunk 1 above)
Comment (fee1-dead) [reply compiler/rustc_typeck/src/check/wfcheck.rs:1527]: This has been done, but only from one producer of wf obligations, to fix this issue. The second fix is unfortunate but is the only solution I can find without breaking the tests.

File: src/test/ui/rfc-2632-const-trait-impl/issue-100222.rs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+#[const_trait]
+pub trait Index {
+    type Output;
+}
+
+#[const_trait]
+pub trait IndexMut where Self: Index {
+    fn foo(&mut self) -> <Self as Index>::Output;
+}
```
Comment (oli-obk) [inline src/test/ui/rfc-2632-const-trait-impl/issue-100222.rs:12]: So... the problem was that this is not `Self: ~const Index`, but it was tested as such? Do we have tests for various combinations of `#[const_trait]` being on the super trait or not, and being on the trait itself or not?

**Comment by oli-obk** [comment]
r=me with CI passing ^^

**Pull request merged**